modules/currency-l1/src/main/scala/org/tessellation/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,64 +51,146 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
+                  .flatMap { estimatedFee =>
+                    EstimatedFee
+                      .getUpdateHash(update, dataApplication.serializeUpdate)
+                      .map(EstimatedFeeResponse(estimatedFee, _))
+                  }
+                  .map(_.asJson.dropNullValues)
+                  .flatMap(Ok(_))
+            }
           }
         }
         .handleUnknownError
 
     case GET -> Root / "data" =>
-      implicit val signedEncoder: Encoder[Signed[DataUpdate]] = Signed.encoder(dataApplication.dataEncoder)
-      dataUpdatesQueue.view.flatMap(Ok(_))
+      implicit val dataUpdateEncoder: Encoder[DataUpdate] = dataApplication.dataEncoder
+      dataTransactionsQueue.view.flatMap(Ok(_))
 
     case req @ POST -> Root / "data" =>
-      implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
-      req
-        .asR[Signed[DataUpdate]] {
-          _.toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _).flatMap {
-            case Left(_) => BadRequest(InvalidSignature.toApplicationError)
-            case Right(hashed) =>
-              validate(hashed.signed) {
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def parseMultipleDataTransactions: F[DataTransactions] = {
+        implicit val dataTransactionRequest: EntityDecoder[F, DataTransactions] = dataApplication.requestDecoder
+        req.as[DataTransactions]
+      }
+
+      def parseSingleDataTransaction: F[Signed[DataUpdate]] = {
+        implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
+        req
+          .as[Signed[DataUpdate]]
+      }
+
+      def handleDataUpdates(dataTransactions: DataTransactions): F[Response[F]] =
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+
+              case feeTransaction: FeeTransaction =>
+                Signed(feeTransaction, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](serialize[F] _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+            }
+          }.flatMap { results =>
+            val errors = results.collect { case Left(_) => InvalidSignature.toApplicationError }
+            val validTransactions = results.collect { case Right(hashed) => hashed }
+
+            if (errors.nonEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("Invalid signature in data transactions"))
```

Comment:
Why do you map `Left` on errors to `InvalidSignature` if you don't use it anyway?
Maybe something like this?

```scala
val (errors, transactions) = results.partitionMap(identity)

NonEmptyList.fromList(transactions) match {
  case None => BadRequest(...).pure[F]
  case Some(txs) => dataTransactionsQueue.offer(txs.map(_.signed)).map(mkResponse)
}

def mkResponse(transactions: NonEmptyList[HashedTransactions]): Response = {
  val (dataUpdateHashes, feeTransactionHashes) = transactions.toList.partition(_.signed.value.isInstanceOf[DataUpdate])
  
  (dataUpdatesHashes, feeTransactionHashes) match {
    case (data, Nil) => Ok("hash" ->> ...)
    case (data, fees) => Ok("hash" ->> [], "feeHash" ->> [])
  }
  
  
}
```

Reply:
btw, if you use `feeHash` not `feeHashes`, I'd use `hash` for an array, not `hashes`.

Reply:
👍🏼